File: frontend/src/lib/api/endpoints/visualization.ts
import { apiPostDirect, apiGetDirect } from '../client';

export interface VisualizationRequest {
  audio_path: string;
  style: 'circular' | 'radial' | 'bars' | 'particle';
  color_scheme: string;
  resolution: '720p' | '1080p' | '4k';
}

export interface VisualizationTask {
  task_id: string;
  status: 'pending' | 'analyzing' | 'rendering' | 'completed' | 'failed';
  progress: number;
  video_path?: string;
  error?: string;
}

export interface VisualizationTaskResponse {
  task_id: string;
  message: string;
}

export const createVisualization = async (data: VisualizationRequest): Promise<VisualizationTaskResponse> => {
  return apiPostDirect<VisualizationTaskResponse, VisualizationRequest>('/visualization/render', data);
};

export const getVisualizationStatus = async (taskId: string): Promise<VisualizationTask> => {
  return apiGetDirect<VisualizationTask>(`/visualization/status/${taskId}`);
};

File: backend/app/schemas/visualization.py
"""
Visualization schemas for audio analysis and video rendering.
"""
from pydantic import BaseModel
from typing import Literal


class AudioFrameData(BaseModel):
    """Single frame of audio analysis data."""
    time: float           # Timestamp in seconds
    amplitude: float      # RMS amplitude (0-1)
    spectrum: list[float] # Spectrum data (128 frequency bands)


class AudioAnalysisData(BaseModel):
    """Complete audio analysis result."""
    duration: float       # Total duration in seconds
    fps: int              # Frame rate (30)
    total_frames: int
    frames: list[AudioFrameData]


class VisualizationRequest(BaseModel):
    """Request to create a visualization video."""
    audio_path: str       # Relative path to audio file (e.g., "ode-to-joy/total.mp3")
    style: Literal["circular", "radial", "bars", "particle"] = "circular"
    color_scheme: str = "blue"
    resolution: Literal["720p", "1080p", "4k"] = "1080p"


class VisualizationTask(BaseModel):
    """Visualization task status."""
    task_id: str
    status: Literal["pending", "analyzing", "rendering", "completed", "failed"]
    progress: int = 0     # 0-100
    video_path: str | None = None
    error: str | None = None


class VisualizationTaskResponse(BaseModel):
    """Response when creating a visualization task."""
    task_id: str
    message: str
